README.md
![versions](docs/images/profiler-pybadge-w-logo.svg) ![tests](https://github.com/wlan-pi/profiler/workflows/tests/badge.svg) ![coverage-badge](coverage.svg) [![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-v2.0%20adopted-ff69b4.svg)](CODE_OF_CONDUCT.md) [![packagecloud-badge](https://img.shields.io/badge/deb-packagecloud.io-844fec.svg)](https://packagecloud.io/)

# wlanpi-profiler

Profiler is a Wi-Fi client capability analyzer tool built for the [WLAN Pi](https://github.com/WLAN-Pi/).

The primary purpose is to automate the collection and analysis of association request frames.

It performs two primary functions:

1. advertises a "fake" Access Point
2. "profiles" any attempted client association requests (which contain the client's claimed capabilities) 

## Why?

Understanding the various client capabilities found in a particular environment helps in the Wireless LAN definition, design, and troubleshooting/validation process.

The WLAN designer may desire to factor in the capabilities of expected clients in their design output. How many spatial streams? What is the client Tx power? What frequency bands does it support? 

The WLAN troubleshooter may understand better the issues they are uncovering when knowing the capabilities of the client. Does the client support .11k/r/v? Which PHYs does the client support? What channels does the client support?

The profiler helps to more quickly answer these questions for you.

## Client capabilities will vary

Capabilities across each client type will vary; depending on factors like client chipset, the number of antennas, power mode (e.g. iOS Low Power Mode), age of the client, driver, supplicant, etc.

Each client includes its capability details in the 802.11 association frame sent from the client to an access point. By capturing this frame, it is possible to decode and report on the client's claimed capabilities.

However, please note that the client will match the capabilities advertised by an access point. For instance, a 3 spatial stream client will tell a 2 spatial stream AP it only supports 2 spatial streams.

The profiler attempts to address this problem by advertising the highest-level feature sets.   

## Getting started

Start the profiler. It will broadcast a fake AP.

When the client attempts association to the profilers fake AP, it will send an association frame. 

The capabilities of a client are determined based on the tool analyzing the association frame. 

If running directly from a terminal, once profiled, a text report prints in real-time to the screen, and results write to a local directory on the WLAN Pi host. 

Saved results include a copy of the text report and the association frame in PCAP format, and the result is appended to a rotating `.csv` report.

1. Start the profiler:

    - Ensure a supported WLAN NIC is plugged into the WLAN Pi

    1. Starting the profiler service using the Front Panel Menu System (FPMS):
        - Navigate to `Menu` > `Apps` > `Profiler` > `Start`

    2. Starting the service manually:
        - `sudo service wlanpi-profiler start|stop|status`
        - Want to view more of the journal scrollback from `service wlanpi-profiler status` output?

            ```
            journalctl -u wlanpi-profiler.service
            ```

    3. Starting from the terminal:
        - `sudo profiler`


    - How to view CLI usage and man page from the shell:

        ```
        profiler -h
        man wlanpi-profiler
        ```

2. Profile the client:

    - once the profiler is started, the configured SSID will broadcast (default: "WLAN Pi")

    - connect a client and enter any random 8 characters for the PSK

    - note the client will expectedly fail authentication but we should receive the association request

3. Viewing the results:

    - You can look on the WebUI (http://<IPv4_of_WLANPi>/profiler) or on the filesystem at `/var/www/html/profiler`.

## Installation

profiler is included in the [WLAN Pi](https://github.com/WLAN-Pi/) image, but if you want to install it manually, here is what you need.

General requirements:

- adapter (and driver) which supports both monitor mode and packet injection
  - mt76x2u (recommended) and rtl88XXau adapters are tested regularly (everything else is experimental and not officially supported)
- elevated permissions

Package requirements:

- Python version 3.7 or higher
- `iw`, `netstat`, and `tcpdump` tools installed on the host

### Manual installation example with pipx (this will change to debian packaging in the near future)

```
sudo -i

# install depends
apt-get install python3 python3-pip python3-venv git iw netstat tcpdump

# install pipx
python3 -m pip install pipx
python3 -m pipx ensurepath

# install profiler from github
pipx install git+https://github.com/WLAN-Pi/profiler.git

# set reg domain (some adapters/drivers require this in order to Tx in 5 GHz bands)
iw reg set US
```

And starting profiler looks like this:

```
sudo profiler
```

Stop with `CTRL + C`.

# Usage

```
usage: profiler [-h] [-c CHANNEL] [-i INTERFACE] [-s SSID]
                [--config FILE] [--files_path PATH] [--hostname_ssid]
                [--logging [{debug,warning}]] [--noprep] [--noAP] [--no11r]
                [--no11ax] [--oui_update] [--read PCAP] [--version]

wlanpi-profiler is an 802.11 client capabilities profiler. Read the manual with: man wlanpi-profiler

optional arguments:
  -h, --help            show this help message and exit
  -c CHANNEL            set the operating channel to broadcast on
  -i INTERFACE          set network interface for profiler
  -s SSID               set profiler SSID name
  --config FILE         customize path for configuration file (default: /etc/wlanpi-profiler/config.ini)
  --files_path PATH     customize default directory where analysis is saved on local system (default: /var/www/html/profiler)
  --hostname_ssid       use the WLAN Pi's hostname as SSID name (default: False)
  --logging [{debug,warning}]
                        change logging output
  --noprep              disable interface preperation (default: False)
  --noAP                enable Rx only mode (default: False)
  --no11r               turn off 802.11r Fast Transition (FT) reporting
  --no11ax              turn off 802.11ax High Efficiency (HE) reporting
  --oui_update          initiates update of OUI database (requires Internet connection)
  --read PCAP           read and analyze association request frames from pcap
  --version, -V         show program's version number and exit
```

## Usage Examples

We require elevated permissions to put the interface in monitor mode and to open raw native sockets for frame injection. Starting and stopping profiler from the WLAN Pi's Front Panel Menu System (FPMS) will handle this for you automatically. 

```
# capture frames on channel 48 using the default SSID
sudo profiler -c 48
```

```
# capture frames on channel 36 using an SSID called 'JOIN ME'
sudo profiler -c 36 -s "JOIN ME"
```

```
# capture frames on channel 100 with 802.11r disabled for clients that don't like 802.11r
sudo profiler -c 100 --no11r
```

```
# capture frames on the default channel with 802.11ax disabled for clients that don't like 802.11ax
sudo profiler --no11ax
```

```
# capture frames on channel 100 without the fake AP running (Rx only, no Tx)
sudo profiler --noAP -c 100
```

```
# analyze an association request in a previously captured PCAP file (must be the only frame in the file)
sudo profiler --read assoc_frame.pcap
```

```
# increase output to screen for debugging
sudo profiler --logging debug
```

## Feature: overriding defaults with configuration file support

To change the default operation of the script (without passing in CLI args), on the WLAN Pi, a configuration file can be found at `/etc/wlanpi-profiler/config.ini`. 

This can be used as a way to modify settings loaded at runtime such as channel, SSID, and interface. 

## Feature: client capabilities diff

When a client is profiled, a hash of the capabilities is calculated and stored in memory. 

If subsequent association requests are seen from the same client, the previously calculated hash is compared to what is already in memory.

If the hash is the same, the additional association request is ignored. 

If the hash is different, capabilities are profiled and a text diff of the client report is saved.

## Feature: MAC OUI database update

A lookup feature is included to show the manufacturer of the client based on the 6-byte MAC OUI. this is a wrapper around a Python module called `manuf` which uses a local flat file for OUI lookup. 

If you find that some clients are not identified in the results, the flat file may need to be updated.

When the WLAN Pi has connectivity to the internet, this can be done from the CLI of the WLAN Pi:

```
sudo profiler --oui_update
```

## Notes and Warnings

- A client will generally only report the capabilities it has that match the network it associates to.
    - If you want the client to report all of its capabilities, it must associate with a network that supports those capabilities (e.g, a 3 spatial stream client will not report it supports 3 streams if the AP it associates with supports only 1 or 2 streams).
    - The fake AP created by the profiler attempts to simulate a fully-featured AP, but there may be cases where it does not behave as expected.
- Treat reporting of 802.11k capabilities with caution. To be sure verify through other means like:
    - Check neighbor report requests from a WLC/AP debug.
    - Gather and analyze a packet capture for action frames containing the neighbor report.
- While we try our best to make this as accurate as possible, we do not guarantee the accuracy of reporting. **Trust, but verify.**

# Thanks 

- Jerry Olla, Nigel Bowden, and the WLAN Pi Community for all their input and effort on the first versions of the profiler. Without them, this project would not exist.

# Contributing

Want to contribute? Thanks! Please take a few moments to [read this](CONTRIBUTING.md).

# Discussions and Issues

Please use GitHub discussions for dialogue around features and ideas that do not exist. Create issues for problems found running profiler.


HISTORY.md
Releaes 1.0.7

- Allow read of pcap file when no WLAN NICs are present
- Allow reading and analysis of multiple association requests in a single pcap
- Make output to screen less noisy by default
- Add output to JSON flat file to support backend
- Fix listen only/passive mode (--noAP) which listens for any association request on a set channel
- Remove draft label from 802.11ax detection
- Add MCS detection for 802.11ac and 802.11ax
- Add 160 MHz support for 802.11ac and 802.11ax
- Add feature detection for 802.11ax TWT, Punctured Preamble, HE SU Beamformer, HE SU Beamformee, HE (ER) Extended Range, UORA (Uplink OFDMA Random Access), and BSR Control (Buffer Status Report)
- Basic Wi-Fi 6E client association pcap profiling
- Add Intel to client manufacturer detection heuristic
- Interface staging improvements
- Fix crash that happens when utf-8 decoding fails

Release 1.0.6

- Replace pipx packaging with debian packaging

Release 1.0.5

- Add support for deleting profiled files (.txt and .pcap) in addition to reports
- Improve test coverage

Release 1.0.4 

- Fix issue preventing a client from being profiled again in the same session when switching between Private MAC and Device MAC
- Fix problem with oui lookups sometimes failing
- Add Samsung to the client heuristics detection
- Add unit tests for client heuristics detection

Release 1.0.3

- Add heuristics to resolve the manufacturer when a client is using a randomized mac address
- Fix forged beacon interval

Release 1.0.2

- Add basic detection for randomized mac addresses

Release 1.0.1

- Bump scapy version from 2.4.3 to 2.4.4

Release 1.0.0-alpha3

- Improve test coverage
- Begin detecting and reporting on changed client capabilities
- App now looks for config.ini in /etc/profiler2/

Release 1.0.0-alpha2

- Add --yes arg to support automatic yes to --clean y/n prompt
- Refactors; bump tox version in requirements
- Remove menu mode code that is no longer used in WLAN Pi v2 builds
- Allow user to check help usage and version without root permissions
- Remove airmon-ng check kill from interface preparation when profiler is run
- Refactor and improve test coverage
- Fix error where argparse fails to allow profiler to run on channel 2
- Rename GH workflow

Release 1.0.0-alpha1

- Bump major to 1 because package is in production WLAN Pi image.
- Create and change default branch from master to main
- Fix handling of `--no11r` and `--no11ax` modes from either CLI args or config file
- Menu mode tweaks and is now enabled by default
- Minor changes to flat file output such as including the capture channel
- `manuf` is no longer abandoned: replaced it's fork `manuf-ng` and updated `manuf` to working version

Release 0.1.dev6:

- Fix a helper function returning wrong IP of SSH session
- Place all flat files for a particular client in its subfolder
- Lock down scapy to version 2.4.3
- Change 5.0 GHz band marking to 5.8GHz

Release 0.1.dev5:

- Fix bug that prevents profiler.service starting from cockpit
- Only run on Python>=3.7 

Release 0.1.dev4:

- Specify required scapy version in install_requires and requirements

Release 0.1.dev3:

- Console_scripts entry point changed from profiler2 to profiler
- Minor bug fixes
- Fix permissions check

Release 0.1.dev2:

- 2.4 or 5 GHz markings are now appended to filenames for flat files
- Minor refactors and bug fixes
- Allow user to run w/o valid WLAN interface for pcap analysis mode (say running from WSL)
- Remove pymongo dependency and related code

Release 0.1.dev1:

- Port of original [profiler](https://github.com/WLAN-Pi/profiler) to Python 3.7 (:snake:)
- Dependency on `sheriffsparks/scapy-fakeap` removed
- Project is now packaged instead of a single script file (e.g. `python3 -m pip install .` from repo root dir works)
- Enable UTF-8 support for SSID broadcasting (because emoji :grinning:)
- BPF filters applied to scapy's `sniff()` function in attempt to address perf issues with processing beacon overhead
- Leverage multiprocessing stdlib package to separate beacon, sniffing, and profiler code onto different processes
- Interface prepartion commands `iwconfig` and `ipconfig` replaced with `iw` and `ip` equivalents
- `manuf` package appears to be abandoned in Py3. replaced with [manuf-ng](https://github.com/daniel-leicht/manuf-ng)
- Add `--oui_update` switch to trigger manuf.py to update the local OUI flat file from Internet
- Add `--no11ax` switch to disable Tx of 802.11ax HE information elements
- Add `--host_ssid` switch to use the WLAN Pi's hostname as the SSID
- Add `--logging` switch to increase print verbosity for debugging
- Begin use of `pytest` and `tox` to standardize testing and linting
- Adopt a license, style guide, contributing guidelines, and code of conduct


profiler/helpers.py
# -* coding: utf-8 -*-
#
# profiler : a Wi-Fi client capability analyzer tool
# Copyright : (c) 2020-2021 Josh Schmelzle
# License : BSD-3-Clause
# Maintainer : josh@joshschmelzle.com

"""
profiler.helpers
~~~~~~~~~~~~~~~~

provides init functions that are used to help setup the app.
"""

# standard library imports
import argparse
import configparser
import inspect
import json
import logging
import logging.config
import os
import shutil
import signal
import socket
import subprocess
import sys
from base64 import b64encode
from dataclasses import dataclass
from distutils.util import strtobool
from time import ctime
from typing import Any, Dict, List, Union

# third party imports
try:
    import manuf
except ModuleNotFoundError as error:
    if error.name == "manuf":
        print("required module manuf not found.")
    else:
        print(f"{error}")
    sys.exit(signal.SIGABRT)


def run_cli_cmd(cmd: List) -> str:
    """Run an arbitrary CLI command and return the results"""
    cp = subprocess.run(
        cmd,
        encoding="utf-8",
        shell=False,
        check=False,
        capture_output=True,
    )

    if cp.stdout:
        return cp.stdout
    if cp.stderr:
        return cp.stderr

    return "completed process return code is non-zero with no stdout or stderr"


# is tcpdump installed?
try:
    result = run_cli_cmd(["tcpdump", "--version"])
except OSError:
    print(
        "WARNING: problem checking tcpdump version. is tcpdump installed and functioning? did you run with elevated permissions?"
    )


# app imports
from .__version__ import __version__
from .constants import CHANNELS, CONFIG_FILE

FILES_PATH = "/var/www/html/profiler"


def setup_logger(args) -> None:
    """Configure and set logging levels"""
    if args.logging:
        if args.logging == "debug":
            logging_level = logging.DEBUG
        if args.logging == "warning":
            logging_level = logging.WARNING
    else:
        logging_level = logging.INFO

    default_logging = {
        "version": 1,
        "disable_existing_loggers": False,
        "formatters": {
            "standard": {"format": "%(asctime)s [%(levelname)s] %(name)s: %(message)s"}
        },
        "handlers": {
            "default": {
                "level": logging_level,
                "formatter": "standard",
                "class": "logging.StreamHandler",
                "stream": "ext://sys.stdout",
            }
        },
        "loggers": {"": {"handlers": ["default"], "level": logging_level}},
    }
    logging.config.dictConfig(default_logging)


def channel(value: str) -> int:
    """Check if channel is valid"""
    ch = int(value)
    if any(ch in band for band in CHANNELS.values()):
        return ch
    raise ValueError("%s is not a valid channel", ch)


def ssid(ssid: str) -> str:
    """Check if SSID is valid"""
    if len(ssid) > 32:
        raise ValueError("%s length is greater than 32" % ssid)
    return ssid


def setup_parser() -> argparse.ArgumentParser:
    """Set default values and handle arg parser"""
    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description="wlanpi-profiler is an 802.11 client capabilities profiler. Read the manual with: man wlanpi-profiler",
    )
    parser.add_argument(
        "--pytest",
        dest="pytest",
        action="store_true",
        default=False,
        help=argparse.SUPPRESS,
    )
    parser.add_argument(
        "-c",
        dest="channel",
        type=channel,
        help="set the channel to broadcast on",
    )
    parser.add_argument(
        "-i",
        dest="interface",
        help="set network interface for profiler",
    )
    ssid_group = parser.add_mutually_exclusive_group()
    ssid_group.add_argument("-s", dest="ssid", type=ssid, help="set profiler SSID name")
    parser.add_argument(
        "--config",
        type=str,
        metavar="FILE",
        default=CONFIG_FILE,
        help="customize path for configuration file (default: %(default)s)",
    )
    parser.add_argument(
        "--files_path",
        metavar="PATH",
        dest="files_path",
        default="/var/www/html/profiler",
        help="customize default directory where analysis is saved on local system (default: %(default)s)",
    )
    ssid_group.add_argument(
        "--hostname_ssid",
        dest="hostname_ssid",
        action="store_true",
        default=False,
        help="use the WLAN Pi's hostname as SSID name (default: %(default)s)",
    )
    parser.add_argument(
        "--logging",
        help="change logging output",
        nargs="?",
        choices=("debug", "warning"),
    )
    parser.add_argument(
        "--noprep",
        dest="no_interface_prep",
        action="store_true",
        default=False,
        help="disable interface preperation (default: %(default)s)",
    )
    ssid_group.add_argument(
        "--noAP",
        dest="listen_only",
        action="store_true",
        default=False,
        help="enable Rx only mode (default: %(default)s)",
    )
    dot11r_group = parser.add_mutually_exclusive_group()
    dot11r_group.add_argument(
        "--11r",
        dest="ft_enabled",
        action="store_true",
        default=False,
        help=argparse.SUPPRESS,  # "turn on 802.11r Fast Transition (FT) reporting (override --config <file>)",
    )
    dot11r_group.add_argument(
        "--no11r",
        dest="ft_disabled",
        action="store_true",
        default=False,
        help="turn off 802.11r Fast Transition (FT) reporting",
    )
    dot11ax_group = parser.add_mutually_exclusive_group()
    dot11ax_group.add_argument(
        "--11ax",
        dest="he_enabled",
        action="store_true",
        default=False,
        help=argparse.SUPPRESS,  # "turn on 802.11ax High Efficiency (HE) reporting (override --config <file>)",
    )
    dot11ax_group.add_argument(
        "--no11ax",
        dest="he_disabled",
        action="store_true",
        default=False,
        help="turn off 802.11ax High Efficiency (HE) reporting",
    )
    parser.add_argument(
        "--no_sniffer_filter",
        dest="no_sniffer_filter",
        action="store_true",
        default=False,
        help=argparse.SUPPRESS,
    )
    parser.add_argument(
        "--clean",
        dest="clean",
        action="store_true",
        default=False,
        help=argparse.SUPPRESS,
    )
    parser.add_argument(
        "--files",
        dest="files",
        action="store_true",
        default=False,
        help=argparse.SUPPRESS,
    )
    parser.add_argument(
        "--yes", dest="yes", action="store_true", default=False, help=argparse.SUPPRESS
    )
    parser.add_argument(
        "--oui_update",
        dest="oui_update",
        action="store_true",
        default=False,
        help="initiates update of OUI database (requires Internet connection)",
    )
    parser.add_argument(
        "--read",
        metavar="PCAP",
        dest="pcap_analysis",
        help="read and analyze association request frames from pcap",
    )
    parser.add_argument("--version", "-V", action="version", version=f"{__version__}")
    return parser


def files_cleanup(directory: str, acknowledged: bool) -> None:
    """Purge files recursively"""
    log = logging.getLogger(inspect.stack()[0][3])

    from pathlib import Path

    result = list(Path(directory).rglob("*"))
    log.warning("Delete the following files: %s", ", ".join([str(x) for x in result]))

    if acknowledged:
        pass
    elif not input("Are you sure? (y/n): ").lower().strip()[:1] == "y":
        sys.exit(1)

    try:
        for _path in os.listdir(Path(directory)):
            path = Path(directory) / Path(_path)
            if path.is_file():
                print(f"Removing file: {path}")
                path.unlink()
            if path.is_dir():
                print(f"Removing directory: {path}")
                shutil.rmtree(path)
    except OSError:
        log.exception("issue removing files")


def setup_config(args):
    """Create the configuration (SSID, channel, interface, etc) for the Profiler"""
    log = logging.getLogger(inspect.stack()[0][3])

    # load in config (a: from default location "/etc/wlanpi-profiler/config.ini" or b: from provided)
    if os.path.isfile(args.config):
        parser = load_config(args.config)
        # we want to work with a dict whether we have config.ini or not
        config = convert_configparser_to_dict(parser)
    else:
        log.warning("can not find config at %s", args.config)
        config = {}

    if "GENERAL" not in config:
        config["GENERAL"] = {}

    if "channel" not in config["GENERAL"]:
        config["GENERAL"]["channel"] = 36

    if "ssid" not in config["GENERAL"]:
        config["GENERAL"]["ssid"] = "WLAN Pi"

    if "interface" not in config["GENERAL"]:
        config["GENERAL"]["interface"] = "wlan0"

    # handle special config.ini settings
    if config["GENERAL"].get("hostname_ssid"):
        config["GENERAL"]["ssid"] = socket.gethostname()

    # handle args
    #  - args passed in take precedent over config.ini values
    #  - did user pass in options that over-ride defaults?
    if args.channel:
        config["GENERAL"]["channel"] = args.channel
    if args.interface:
        config["GENERAL"]["interface"] = args.interface
    if args.ssid:
        config["GENERAL"]["ssid"] = args.ssid
    elif args.hostname_ssid:
        config["GENERAL"]["ssid"] = socket.gethostname()
    if args.ft_enabled:
        config["GENERAL"]["ft_disabled"] = False
    if args.ft_disabled:
        config["GENERAL"]["ft_disabled"] = args.ft_disabled
    if args.he_enabled:
        config["GENERAL"]["he_disabled"] = False
    if args.he_disabled:
        config["GENERAL"]["he_disabled"] = args.he_disabled
    if args.listen_only:
        config["GENERAL"]["listen_only"] = args.listen_only
    if args.pcap_analysis:
        config["GENERAL"]["pcap_analysis"] = args.pcap_analysis
    if args.files_path:
        config["GENERAL"]["files_path"] = args.files_path
    else:
        config["GENERAL"]["files_path"] = FILES_PATH

    # ensure channel 1 is an integer and not a bool
    try:
        ch = config.get("GENERAL").get("channel")
        if ch:
            ch = int(ch)
        config["GENERAL"]["channel"] = ch
    except KeyError:
        log.warning("config.ini does not have channel defined")

    return config


def convert_configparser_to_dict(config: configparser.ConfigParser) -> Dict:
    """
    Convert ConfigParser object to dictionary.

    The resulting dictionary has sections as keys which point to a dict of the
    section options as key => value pairs.

    If there is a string representation of truth, it is converted from str to bool.
    """
    _dict: "Dict[str, Any]" = {}
    for section in config.sections():
        _dict[section] = {}
        for key, _value in config.items(section):
            try:
                _value = bool(strtobool(_value))
            except ValueError:
                pass
            _dict[section][key] = _value
    return _dict


def load_config(config_file: str) -> configparser.ConfigParser:
    """Load in config from external file"""
    config = configparser.ConfigParser()
    config.read(config_file)
    return config


def validate(config) -> bool:
    """Validate minimum config to run is OK"""
    log = logging.getLogger(inspect.stack()[0][3])

    if not check_config_missing(config):
        return False

    try:
        _ssid = config.get("GENERAL").get("ssid")
        if _ssid:
            ssid(_ssid)

        ch = config.get("GENERAL").get("channel")
        if ch:
            channel(ch)

        verify_reporting_directories(config)
    except ValueError:
        log.error("%s", sys.exc_info()[1])
        sys.exit(signal.SIGABRT)

    return True


def is_randomized(mac) -> bool:
    """Check if MAC Address <format>:'00:00:00:00:00:00' is locally assigned"""
    return any(local == mac.lower()[1] for local in ["2", "6", "a", "e"])


def check_config_missing(config: Dict) -> bool:
    """Check that the minimal config items exist"""
    log = logging.getLogger(inspect.stack()[0][3])
    try:
        if "GENERAL" not in config:
            raise KeyError("missing general section from configuration")
        options = config["GENERAL"].keys()
        if "interface" not in options:
            raise KeyError("missing interface from config")
        if "channel" not in options:
            raise KeyError("missing channel from config")
        if "ssid" not in options:
            raise KeyError("missing ssid from config")

    except KeyError:
        log.error("%s", sys.exc_info()[1])
        return False
    return True


def update_manuf() -> bool:
    """Manuf wrapper to update manuf OUI flat file from Internet"""
    log = logging.getLogger(inspect.stack()[0][3])
    try:
        flat_file = os.path.join(manuf.__path__[0], "manuf")
        manuf_location = f"{sys.prefix}/bin/manuf"
        log.info("OUI database is located at %s", flat_file)
        log.info("manuf is located at %s", manuf_location)
        log.info(
            "manuf file last modified at: %s",
            ctime(os.path.getmtime(flat_file)),
        )
        log.info("running 'sudo manuf --update'")
        out = run_cli_cmd(["sudo", manuf_location, "--update"])
        log.info("%s", str(out))
        if "URLError" not in out:
            log.info(
                "manuf file last modified at: %s",
                ctime(os.path.getmtime(flat_file)),
            )
    except OSError:
        log.exception("problem updating manuf. make sure manuf is installed...")
        print("exiting...")
        return False
    return True


def verify_reporting_directories(config: Dict) -> None:
    """Check reporting directories exist and create if not"""
    log = logging.getLogger(inspect.stack()[0][3])

    if "GENERAL" in config:
        files_path = config["GENERAL"].get("files_path")
        if not os.path.isdir(files_path):
            log.debug(os.makedirs(files_path))

        clients_dir = os.path.join(files_path, "clients")

        if not os.path.isdir(clients_dir):
            log.debug(os.makedirs(clients_dir))

        reports_dir = os.path.join(files_path, "reports")

        if not os.path.isdir(reports_dir):
            log.debug(os.makedirs(reports_dir))


def get_frequency_bytes(channel: int) -> bytes:
    """Take a channel number, converts it to a frequency, and finally to bytes"""
    if channel == 14:
        freq = 2484
    if channel < 14:
        freq = 2407 + (channel * 5)
    elif channel > 14:
        freq = 5000 + (channel * 5)

    return freq.to_bytes(2, byteorder="little")


class Base64Encoder(json.JSONEncoder):
    """A Base64 encoder for JSON"""

    # example usage: json.dumps(bytes(frame), cls=Base64Encoder)

    # pylint: disable=method-hidden
    def default(self, obj):
        """Perform default Base64 encode"""
        if isinstance(obj, bytes):
            return b64encode(obj).decode()
        return json.JSONEncoder.default(self, obj)


def get_wlanpi_version() -> str:
    """Retrieve system image verson"""
    wlanpi_version = "unknown"
    try:
        with open("/etc/wlanpi-release") as _file:
            lines = _file.read().splitlines()
            for line in lines:
                if "VERSION" in line:
                    wlanpi_version = "{0}".format(
                        line.split("=")[1].replace('"', "").replace("'", "").strip()
                    )
    except OSError:
        pass
    return wlanpi_version


def flag_last_object(seq):
    """Treat the last object in an iterable differently"""
    seq = iter(seq)  # ensure seq is an iterator
    _a = next(seq)
    for _b in seq:
        yield _a, False
        _a = _b
    yield _a, True


def generate_run_message(config: Dict) -> None:
    """Create message to display to users screen"""
    if config["GENERAL"].get("listen_only") is True:
        out = []
        out.append("Starting profiler in listen only mode:")
        out.append(
            f" - Listening for association frames with {config['GENERAL']['interface']} on channel {config['GENERAL']['channel']}"
        )
        out.append(" - Results are saved locally and printed to screen")
        out.append(" ")
        out.append("Instructions:")
        out.append(
            f" - Associate your Wi-Fi client to any AP on channel {config['GENERAL']['channel']}"
        )
        out.append(" - We should passively detect the association request")
        header_len = len(max(out, key=len))

        print(f"\n{'~' * header_len}")
        for line in out:
            print(line)
        print(f"{'~' * header_len}\n")
    else:
        out = []
        channel = config["GENERAL"]["channel"]
        interface = config["GENERAL"]["interface"]
        ssid = config["GENERAL"]["ssid"]
        if channel:
            out.append(
                f"Starting profiler AP with interface {interface} on channel {channel}:"
            )
        else:
            out.append(f"Starting profiler AP with interface {interface}")
        out.append(f" - Our fake AP SSID: {ssid}")
        out.append(" - Results are saved locally and printed to screen")
        out.append(" ")
        out.append("Instructions:")
        out.append(f" - Associate your Wi-Fi client to our SSID: {ssid}")
        out.append(" - Enter any random password to connect")
        out.append(" ")
        out.append(
            "The client will fail authentication but should send an association request"
        )
        header_len = len(max(out, key=len))

        print(f"\n{'=' * header_len}")
        for line in out:
            print(line)
        print(f"{'=' * header_len}\n")


@dataclass
class Capability:
    """Define custom fields for reporting"""

    name: str = ""
    value: Union[str, int] = ""
    db_key: str = ""
    db_value: Union[int, str, List[str]] = 0


def get_bit(byteval, index) -> bool:
    """Retrieve bit value from byte at provided index"""
    return (byteval & (1 << index)) != 0
